### ISSUE TITLE
Support React 18 concurrent rendering with React's new hook `useSyncExternalStore`

### ISSUE THREAD

**Issue by SukkaW** [nanostores/react#6]
What is React concurrent rendering?

- https://reactjs.org/blog/2022/03/29/react-v18.html
- https://reactjs.org/blog/2022/03/08/react-18-upgrade-guide.html
- https://reactjs.org/blog/2021/12/17/react-conf-2021-recap.html

What is `useSyncExternalStore`?

- https://github.com/reactwg/react-18/discussions/70
- https://github.com/reactwg/react-18/discussions/86

In short, with the new hook `useSyncExternalStore`, the state management library maintainers do not need to write the `forceUpdate` anymore. Instead, React can now subscribe to the external store and trigger a re-render if needed.

I have just implement a PoC of `@nanostores/react` re-written in useSyncExternalStore:

```js
export function useStore(store, opts = {}) {
  if (process.env.NODE_ENV !== 'production') {
    if (typeof store === 'function') {
      throw new Error(
        'Use useStore(Template(id)) or useSync() ' +
        'from @logux/client/react for templates'
      )
    }
  }

  let sub = useCallback(
    (onChange) => opts.keys
      ? listenKeys(
        store,
        opts.keys,
        onChange
      )
      : store.listen(onChange),
    [opts.keys, store]
  )

  let get = store.get.bind(store);

  return useSyncExternalStore(
    sub,
    get, // getClientSideSnapshot
    get // getServerSideSnapshot, while the nanostore can always initiate a value when creating the atom
  )
}
```

Note: Although `useSyncExternalStore` is introduced in React 18, the React Team also publishes a polyfill backporting this hook to support any React version with hooks (which is `> 16.8`)

**Comment by ai** [comment]
Looks very interesting. What is benefits for end users?

**Comment by YeungKC** [comment]
```typescript
nanostore(1);
nanostore.set(1);
```

In the current behavior, this code would be call to `forceRender`, but value no change.


https://github.com/nanostores/react/blob/main/index.js#L23

**Comment by SukkaW** [comment]
> Looks very interesting. What is benefits for end users?

Unlike `forceRender` which is not supported by React, useSyncExternalStore is React official, first-party built-in hooks that designed for external store library.
Since useSyncExternalStore is provided by React, it has the better compatiblity with React, and will not break React concurrent rendering in the future.

But it has downside as well. useSyncExternalStore is not available on React 16 and React 17. Although the React Team does provide a polyfill, it still means for React 16/17 users, the package size will increase by about 500 bytes.

**Comment by ai** [comment]
It is OK to release a major version with React 18 requirement.

But we need a rational reason to do it, not just having the same feature in official API.

Will we have a better performance as @YeungKC told? If so send a PR.

**Comment by SukkaW** [comment]
> But we need a rational reason to do it, not just having the same feature in official API.
> 
> Will we have a better performance as @YeungKC told? If so send a PR.

Yes, we will have a better performance, but not the way @YeungKC described.

Consider this example:

```jsx
const [input, setInput] = useState('');
const searchQuery = useStore(searhQueryStore);
const [isPending, startTransition] = useTransition();

const handleUserInput = useCallback(() => {
  startTransition(() => {
    searhQueryStore.set(newValue);
  });
}, [store]);
```

In this example, `input` is a React internal state. I use it to "control" an input component (make it a "controlled component"). And updating a search query is expensive (searching, filtering, updating the entire search result list). So I wrap it inside React 18's `startTransition` to let React only update `searchQueryStore` when idle (low priority).

Currently, `forceRender` relies on `setState({})`, which is not synchronized, as I quote:

> When you change React state, React doesn’t immediately change the state. Instead, React queues the update and schedules a render. When React starts to render, it will then look at the entire queue of updates, and use different heuristics and algorithms to determine the next update to work on. This process has safeguards and semantics in place to ensure that rendering is always consistent.

So when the `searhQueryStore.set` is called, the React should have started a re-render immediately, but it doesn't. Instead, it waits until `forceUpdate`'s `setState` is scheduled. It results in tearing and mismatching:

![](https://user-images.githubusercontent.com/490574/61502196-ce109200-aa0d-11e9-9efc-6203545d367c.gif)

> Here is a tearing example of React Redux v7 which uses the same `forceRender` approach as shown [here](https://github.com/reduxjs/react-redux/blob/2d3309d3c1378774e157a6238cd2b57019e75b77/src/hooks/useSelector.js#L84)
> React Redux v8 migrates to the `useSyncExternalStore`, and no tearing anymore.

By using `useSyncExternalStore`, an external store (like nanostores) can tell React explicitly that "Do not commit the current in-consistent UI to the DOM, and do a re-render with newest value, right now!".

**Comment by ai** [comment]
Good point, thanks for good explanation

**Comment by ai** [comment]
Released in 0.3. Thanks to @SukkaW for PR.
